crest/Assets/Crest/Crest/Scripts/Helpers/FloatingOrigin.cs

```diff
@@ -65,35 +69,149 @@ public class FloatingOrigin : MonoBehaviour
         [Tooltip("Optionally provide a list of Gerstner components to avoid doing a FindObjectsOfType() call."), SerializeField]
         ShapeGerstnerBatched[] _overrideGerstnerList = null;
 
+        [Space(10)]
+
+        [SerializeField]
+        internal DebugFields _debug = new DebugFields();
+
+        [System.Serializable]
+        internal partial class DebugFields
+        {
+            [Tooltip("GPU capture using external tool on origin shift.")]
+            public bool _captureOnShift;
+            [Tooltip("Pause editor on origin shift.")]
+            public bool _pauseOnShift;
+            [Tooltip("Pause editor before origin shift. When it meets the threshold it postpones the shift to the next frame and pauses this frame.")]
+            public bool _pauseBeforeShift;
+            [Tooltip("Log to console on origin shift.")]
+            public bool _logOnShift;
+            internal bool _isCapturing;
+            internal bool _shiftNextX;
+            internal bool _shiftNextY;
+        }
+
+        bool _isFirstFixedUpdate = true;
+
         ParticleSystem.Particle[] _particleBuffer = null;
 
-        static readonly int sp_CrestFloatingOriginOffset = Shader.PropertyToID("_CrestFloatingOriginOffset");
+        public static readonly int sp_CrestFloatingOriginOffset = Shader.PropertyToID("_CrestFloatingOriginOffset");
 
         Vector3 _originOffset;
 
         public static Vector3 TeleportOriginThisFrame { get; private set; }
 
-        void LateUpdate()
+        IEnumerator Start()
         {
-            TeleportOriginThisFrame = Vector3.zero;
+            while (true)
+            {
+                // NOTE: Will not work in batch mode:
+                // https://docs.unity3d.com/Manual/CLIBatchmodeCoroutines.html
+                yield return Helpers.WaitForEndOfFrame;
+
+                if (_debug._isCapturing)
+                {
+                    ExternalGPUProfiler.EndGPUCapture();
+                    _debug._isCapturing = false;
+                }
+
+#if UNITY_EDITOR
+                if (_debug._pauseOnShift && TeleportOriginThisFrame != Vector3.zero)
+                {
+                    UnityEditor.EditorApplication.isPaused = true;
+                }
+#endif
+
+                // Cannot clear at start of FixedUpdate (without extra logic) as it can run multiple times.
+                TeleportOriginThisFrame = Vector3.zero;
 
+                _isFirstFixedUpdate = true;
+            }
+        }
+
+        void FixedUpdate()
+        {
+            if (!_isFirstFixedUpdate)
```

Comment:
the way i've typically handled this is something like

int _lastUpdateFrame;

void FixedUpdate()
{
    if( _lastUpdateFrame == Time.frameCount )
    {
        return;
    }
    _lastUpdateFrame = Time.frameCount;

    ....
}

this would be a bit simpler for me. also if this were to mean we could make the coroutine above debug-only (so not in final builds) i'd be in favour - but only because i'm hopelessly bad at coroutines and view them with suspicion :). but if you prefer to keep it the way it is and the coroutine is not generating per frame garbage then i wouldnt block the review on it.

Reply:
Thanks. I have changed to your approach as it does reduce our dependency on the coroutine. But we still depend on it because TeleportOriginThisFrame and HasTeleportedThisFrame has to be cleared but FixedUpdate does not run each frame.

Reply:
As for garbage, I did not see any from this. Found some other garbage :\ which I will fix separately.